### PR TITLE
socket-util: Introduce send_one_fd_iov() and receive_one_fd_iov()

### PR DESCRIPTION
**Pull request by filbranden**
I wrote these while working on a way to lock the passwd file through a separate process. (That is not ready yet, but these are useful on their own, so I thought I'd send them now.)

The new functions take a `struct iovec` to send/receive data together with the passed FD.
    
The receive function returns the FD through an output argument. In case data is received, but no FD is passed, the receive function will set the output argument to -1 explicitly.
    
Update code in `dynamic-user.c` to use the new helpers.

Also add tests for passing/receiving FDs, for trying to pass FDs but receiving data only. While at it, fix the last test in the series, which forked a child but was missing an `_exit()` statement (which got exposed when more tests were added.)

### CODE REVIEW

File: src/core/dynamic-user.c [hunk 1]
```diff
@@ -352,42 +334,11 @@ static int dynamic_user_pop(DynamicUser *d, uid_t *ret_uid, int *ret_lock_fd) {
...
-                .msg_iov = &iov,
-                .msg_iovlen = 1,
-        };
-        ssize_t k;
 
         assert(d);
 
         /* Store the UID and lock_fd in the storage socket. This should be called with the socket pair lock taken. */
-
-        if (lock_fd >= 0) {
```
Comment (keszybz) [inline src/core/dynamic-user.c:371]: This checked if `lock_fd` is non-negative, but is replaced with a call to `send_one_fd_iov`, which calls `send_one_fd_iov_sa`, which attaches `fd` unconditionally. Is this on purpose?

Comment (filbranden) [reply src/core/dynamic-user.c:371]: Ah indeed this is a problem and I didn't notice it.

Right now `send_one_fd_iov()` only works when `fd >= 0` (it asserts that), so this would have to branch and use a bare `send()` if there's no FD involved.

But I'm thinking of updating `send_one_fd_iov()` so it needs either FD *or* data (which is similar to how `receive_one_fd_iov()`, which will return >= 0 if any of those are true.) I'll update this PR with that.

Other than this, what do you think of this PR in general?

Comment (filbranden) [reply src/core/dynamic-user.c:371]: This is addressed now. `send_one_fd_iov()` can now send data with no FD attached. (But it needs to send at least one of those, otherwise it's an error.)

I added two more tests for this case.

There's one corner case here that's kind of odd... If you send a message with an empty string in it, it's considered a valid send. On the receive side though, it's hard to distinguish from receiving no FD and no data, and I'm currently returning -EIO. Should that maybe return 0 instead, signaling that no data was received?

PTAL.

**Comment by filbranden** [comment]
Please take a look. I believe this should be ready to merge. Thanks!

File: src/basic/socket-util.c [hunk 2]
```diff
@@ -1016,39 +1017,75 @@ int send_one_fd_sa(
...
+                struct cmsghdr *cmsg;
 
-        return 0;
+                cmsg = CMSG_FIRSTHDR(&mh);
+                cmsg->cmsg_level = SOL_SOCKET;
+                cmsg->cmsg_type = SCM_RIGHTS;
+                cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+                memcpy(CMSG_DATA(cmsg), &fd, sizeof(int));
+
+                mh.msg_controllen = CMSG_SPACE(sizeof(int));
```
Comment (keszybz) [inline src/basic/socket-util.c:1046]: Just set `msg_control` and `msg_controllen` here.

File: src/basic/socket-util.c [hunk 3]
```diff
@@ -1016,39 +1017,75 @@ int send_one_fd_sa(
         struct msghdr mh = {
                 .msg_name = (struct sockaddr*) sa,
                 .msg_namelen = len,
+                .msg_iov = iov,
+                .msg_iovlen = iovlen,
                 .msg_control = &control,
                 .msg_controllen = sizeof(control),
```
Comment (keszybz) [inline src/basic/socket-util.c]: So `msg_control` and `len` are initialized here, but reset below. Please remove the initialization (to let it be zero), and not set it below if not necessary. (I see that this is copied from existing code... But it should be fixed when anyway.)

Comment (filbranden) [reply src/basic/socket-util.c]: Done. Yeah, this looks better without the else branch.

I still need to set msg_controllen twice, since the CMSG_FIRSTHDR() macro depends on it, but for the case where there's no FD it won't be set at all.

Thanks!

Comment (filbranden) [reply src/basic/socket-util.c]: This was addressed in latest version.

**Comment by filbranden** [comment]
All comments addressed... Please take another look.
